packages/protocol/contracts/L1/libs/LibProposing.sol

```diff
@@ -71,28 +99,57 @@ library LibProposing {
         meta.depositsProcessed =
             LibEthDepositing.processDeposits(state, config, input.beneficiary);
 
-        unchecked {
-            meta.timestamp = uint64(block.timestamp);
-            meta.l1Height = uint64(block.number - 1);
-            meta.l1Hash = blockhash(block.number - 1);
-            meta.mixHash = bytes32(block.difficulty * state.numBlocks);
-        }
-
+        // Init the block
         TaikoData.Block storage blk =
-            state.blocks[state.numBlocks % config.ringBufferSize];
+            state.blocks[state.numBlocks % config.blockRingBufferSize];
 
+        blk.metaHash = LibUtils.hashMetadata(meta);
         blk.blockId = state.numBlocks;
-        blk.proposedAt = meta.timestamp;
+        blk.gasLimit = meta.gasLimit;
         blk.nextForkChoiceId = 1;
         blk.verifiedForkChoiceId = 0;
-        blk.metaHash = LibUtils.hashMetadata(meta);
+        blk.proverReleased = false;
+
         blk.proposer = msg.sender;
+        blk.feePerGas = state.feePerGas;
+        blk.proposedAt = meta.timestamp;
 
-        uint64 blockFee;
+        if (assignedProver == address(0)) {
+            if (state.numOpenBlocks >= config.rewardOpenMaxCount) {
+                revert L1_TOO_MANY_OPEN_BLOCKS();
+            }
+            blk.rewardPerGas = state.feePerGas;
+            ++state.numOpenBlocks;
+        } else {
+            blk.assignedProver = assignedProver;
+
+            // Cap the reward to a range of [95%, 105%] * blk.feePerGas, if
+            // rewardPerGasRange is set to 5% (500 bp)
+            uint32 diff = blk.feePerGas * config.rewardPerGasRange / 10_000;
+            blk.rewardPerGas = uint32(
+                uint256(rewardPerGas).min(state.feePerGas + diff).max(
+                    state.feePerGas - diff
+                )
+            );
```

Comment:
As the prover has at least 95% reward, the prover may use a very small rewardpergas value to gain bigger weight?

Reply:
We can adjust the value to 90% or even lower but we do not know what's the best value, so now 95% is used to start testing it.

There is a feature not implemented yet: https://github.com/taikoxyz/taiko-mono/issues/14090